File: 3058-maximum-number-of-k-divisible-components/maximum-number-of-k-divisible-components.py
class Solution:
    def maxKDivisibleComponents(self, n: int, edges: List[List[int]], values: List[int], k: int) -> int:
        answer = 0

        visited = set()
        
        def dfs(node):
            nonlocal answer
            is_leaf = True
            ret = values[node]
            for _next in tree[node]:
                if _next in visited:
                    continue
                visited.add(_next)
                is_leaf = False
                ret += dfs(_next)

            if ret % k == 0:
                answer += 1
                return 0

            return ret

        tree = [[] for _ in range(n)]

        for u, v in edges:
            tree[u].append(v)
            tree[v].append(u)

        visited.add(0)
        dfs(0)
        return answer
        

        
        

File: 3058-maximum-number-of-k-divisible-components/maximum-number-of-k-divisible-components.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number[]} values
 * @param {number} k
 * @return {number}
 */
var maxKDivisibleComponents = function(n, edges, values, k) {
    const graph = Array.from({length: n}, () => []);
    for (const [u, v] of edges) {
        graph[u].push(v);
        graph[v].push(u);
    }

    let answer = 0;
    const dfs = (node, parent) => {
        let sum = 0;

        for (const next of graph[node]) {
            if (next === parent) continue;
            sum += dfs(next, node);
            sum %= k;
        }

        sum += values[node];
        sum %= k;

        if (sum === 0) {
            answer += 1;
        }

        return sum;
    }
    dfs(0, -1);
    return answer;
};